sockeye/data_io.py

```diff
@@ -322,60 +322,72 @@ def shard_data(source_fname: str,
     """
+    source_fname, *source_factor_fnames = source_fnames
+    source_vocab, *source_factor_vocabs = source_vocabs
+
     os.makedirs(output_prefix, exist_ok=True)
     source_shard_fnames = [os.path.join(output_prefix, C.SHARD_SOURCE % i)
                            for i in range(num_shards)]  # type: List[str]
     target_shard_fnames = [os.path.join(output_prefix, C.SHARD_TARGET % i)
                            for i in range(num_shards)]  # type: List[str]
+    source_factor_shard_fnames = [[os.path.join(output_prefix, C.SHARD_SOURCE % j) + ".factor.%d" % i
+                                   for j in range(num_shards)]
+                                  for i in range(len(source_factor_fnames))]  # type: List[List[str]]
 
-    data_stats_accumulator = DataStatisticsAccumulator(buckets, vocab_source, vocab_target,
+    data_stats_accumulator = DataStatisticsAccumulator(buckets, source_vocab, target_vocab,
                                                        length_ratio_mean, length_ratio_std)
-    per_shard_stat_accumulators = [DataStatisticsAccumulator(buckets, vocab_source, vocab_target, length_ratio_mean,
+    per_shard_stat_accumulators = [DataStatisticsAccumulator(buckets, source_vocab, target_vocab, length_ratio_mean,
                                                              length_ratio_std) for shard_idx in range(num_shards)]
 
     with ExitStack() as exit_stack:
-        source_shards = []
-        target_shards = []
-        # create shard files:
-        for fname in source_shard_fnames:
-            source_shards.append(exit_stack.enter_context(smart_open(fname, mode="wt")))
-        for fname in target_shard_fnames:
-            target_shards.append(exit_stack.enter_context(smart_open(fname, mode="wt")))
-        shards = list(zip(source_shards, target_shards, per_shard_stat_accumulators))
-
-        source_iter = SequenceReader(source_fname, vocab_source, add_bos=False)
-        target_iter = SequenceReader(target_fname, vocab_target, add_bos=True)
-
-        random_shard_iter = iter(lambda: random.choice(shards), None)
-
-        for source, target, (source_shard, target_shard, shard_stats) in zip(source_iter, target_iter,
-                                                                             random_shard_iter):
+        source_shards = [exit_stack.enter_context(smart_open(f, mode="wt")) for f in source_shard_fnames]
+        target_shards = [exit_stack.enter_context(smart_open(f, mode="wt")) for f in target_shard_fnames]
+        source_factor_shards = [
+            [exit_stack.enter_context(smart_open(f, mode="wt")) for f in source_factor_shard_fnames[i]] for i in
+            range(len(source_factor_fnames))]
+
+        source_iter = SequenceReader(source_fname, source_vocab, add_bos=False)
+        target_iter = SequenceReader(target_fname, target_vocab, add_bos=True)
+        source_factor_iters = [SequenceReader(f, v, add_bos=False) for f, v in zip(source_factor_fnames,
```

Comment:
nitpicky: maybe spell out f and v as factor and vocab.

Reply:
makes sense